File: src/transform/projection.ts
/**
 * Coordinate Projection using proj4
 *
 * Handles conversion between projected coordinate systems (UTM, State Plane, etc.)
 * and WGS84 geographic coordinates.
 */

import proj4 from 'proj4';
import { GeoreferenceError } from '../utils/errors.js';
import { createLogger } from '../utils/logger.js';
import type { Georeference, WGS84Position } from '../types.js';

const logger = createLogger('projection');

// WGS84 definition (EPSG:4326)
const WGS84 = 'EPSG:4326';

// Common EPSG definitions that proj4 might not have built-in
const EPSG_DEFINITIONS: Record<number, string> = {
  // UTM Zones (WGS84)
  32601: '+proj=utm +zone=1 +datum=WGS84 +units=m +no_defs',
  32636: '+proj=utm +zone=36 +datum=WGS84 +units=m +no_defs',
  32637: '+proj=utm +zone=37 +datum=WGS84 +units=m +no_defs',
  32638: '+proj=utm +zone=38 +datum=WGS84 +units=m +no_defs',

  // Add more as needed...
  // State Plane coordinates, national grids, etc.
};

/**
 * Get proj4 definition string for an EPSG code
 */
function getEPSGDefinition(epsg: number): string {
  // Check our custom definitions first
  if (EPSG_DEFINITIONS[epsg]) {
    return EPSG_DEFINITIONS[epsg];
  }

  // Try to use proj4's built-in definitions
  const epsgString = `EPSG:${epsg}`;
  try {
    // proj4 will throw if not found
    proj4(epsgString);
    return epsgString;
  } catch {
    // Try to construct common projections
    if (epsg >= 32601 && epsg <= 32660) {
      // UTM North zones
      const zone = epsg - 32600;
      return `+proj=utm +zone=${zone} +datum=WGS84 +units=m +no_defs`;
    }
    if (epsg >= 32701 && epsg <= 32760) {
      // UTM South zones
      const zone = epsg - 32700;
      return `+proj=utm +zone=${zone} +south +datum=WGS84 +units=m +no_defs`;
    }

    throw new GeoreferenceError(
      `Unknown EPSG code: ${epsg}. Please add the proj4 definition.`,
      { epsg }
    );
  }
}

/**
 * Register a custom EPSG definition
 */
export function registerEPSG(epsg: number, definition: string): void {
  EPSG_DEFINITIONS[epsg] = definition;
  proj4.defs(`EPSG:${epsg}`, definition);
  logger.debug({ epsg, definition }, 'Registered custom EPSG definition');
}

/**
 * Convert projected coordinates (Easting/Northing) to WGS84 (lon/lat)
 */
export function projectToWGS84(
  easting: number,
  northing: number,
  epsg: number
): { longitude: number; latitude: number } {
  const sourceProj = getEPSGDefinition(epsg);

  try {
    const [longitude, latitude] = proj4(sourceProj, WGS84, [easting, northing]);

    logger.debug(
      { easting, northing, epsg, longitude, latitude },
      'Projected to WGS84'
    );

    return { longitude, latitude };
  } catch (error) {
    throw new GeoreferenceError(
      `Failed to project coordinates from EPSG:${epsg} to WGS84`,
      { easting, northing, epsg, error }
    );
  }
}

/**
 * Convert WGS84 (lon/lat) to projected coordinates (Easting/Northing)
 */
export function projectFromWGS84(
  longitude: number,
  latitude: number,
  epsg: number
): { easting: number; northing: number } {
  const targetProj = getEPSGDefinition(epsg);

  try {
    const [easting, northing] = proj4(WGS84, targetProj, [longitude, latitude]);
    return { easting, northing };
  } catch (error) {
    throw new GeoreferenceError(
      `Failed to project coordinates from WGS84 to EPSG:${epsg}`,
      { longitude, latitude, epsg, error }
    );
  }
}

/**
 * Convert georeference origin to WGS84 position
 */
export function georeferenceToWGS84(georef: Georeference): WGS84Position {
  const { longitude, latitude } = projectToWGS84(
    georef.origin.easting,
    georef.origin.northing,
    georef.epsg
  );

  // For now, we assume elevation is ellipsoidal height
  // TODO: Add geoid model support for orthometric heights
  let height = georef.origin.elevation;

  if (georef.heightReference === 'orthometric') {
    logger.warn(
      'Orthometric height reference specified but geoid model not loaded. ' +
      'Using elevation directly as ellipsoidal height.'
    );
    // In the future, we could use a geoid model (EGM96, etc.) to convert
    // height = elevation + geoidHeight(latitude, longitude)
  }

  return { longitude, latitude, height };
}

/**
 * Calculate the convergence angle (grid north vs true north) for a projected point
 * This is the angle between grid north and true north at a specific location.
 *
 * For UTM, this can be significant away from the central meridian.
 *
 * Note: The user-provided trueNorthRotation should already account for this,
 * but this function can be used to verify or calculate it.
 */
export function calculateConvergenceAngle(
  easting: number,
  northing: number,
  epsg: number
): number {
  // Convert the point to WGS84
  const { longitude, latitude } = projectToWGS84(easting, northing, epsg);

  // Calculate a small offset to the north in WGS84
  const deltaLat = 0.0001; // ~11 meters
  const northPoint = projectFromWGS84(longitude, latitude + deltaLat, epsg);

  // The angle of the grid line from the original point to the north point
  // relative to grid north (positive Y direction)
  const dE = northPoint.easting - easting;
  const dN = northPoint.northing - northing;

  // Convergence angle in degrees (positive = grid north is west of true north)
  const convergenceAngle = Math.atan2(dE, dN) * (180 / Math.PI);

  logger.debug(
    { easting, northing, epsg, convergenceAngle },
    'Calculated convergence angle'
  );

  return convergenceAngle;
}

File: src/transform/projection.rs
use crate::error::{PhotoTilerError, Result};

/// Project an (easting, northing) pair from the given EPSG CRS to WGS84.
///
/// Returns `(longitude, latitude)` in degrees.
pub fn project_to_wgs84(epsg: u32, easting: f64, northing: f64) -> Result<(f64, f64)> {
    let from = format!("EPSG:{epsg}");
    let proj = proj::Proj::new_known_crs(&from, "EPSG:4326", None).map_err(|e| {
        PhotoTilerError::Transform(format!(
            "Failed to create projection from {from} to WGS84: {e}"
        ))
    })?;

    let (lon, lat) = proj
        .convert((easting, northing))
        .map_err(|e| PhotoTilerError::Transform(format!("Projection failed: {e}")))?;

    Ok((lon, lat))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn utm_zone_36n_to_wgs84() {
        // EPSG:32636 = UTM zone 36N
        // UTM zone 36N central meridian = 33°E
        // (500000, 0) in UTM 36N should map to approximately (33°, 0°)
        let (lon, lat) = project_to_wgs84(32636, 500_000.0, 0.0).unwrap();
        assert!(
            (lon - 33.0).abs() < 0.01,
            "longitude {lon} should be near 33.0"
        );
        assert!(lat.abs() < 0.01, "latitude {lat} should be near 0.0");
    }

    #[test]
    fn utm_zone_36n_known_offset() {
        // The actual test data has offset.xyz: 772598.000 3575069.000 641.000
        // This is a point in UTM zone 36N (EPSG:32636)
        let (lon, lat) = project_to_wgs84(32636, 772_598.0, 3_575_069.0).unwrap();
        // Should be somewhere in the eastern Mediterranean / Middle East
        assert!(lon > 30.0 && lon < 40.0, "longitude {lon} out of range");
        assert!(lat > 30.0 && lat < 35.0, "latitude {lat} out of range");
    }

    #[test]
    fn invalid_epsg_returns_error() {
        let result = project_to_wgs84(99999, 0.0, 0.0);
        assert!(result.is_err());
    }
}
